src/esm_parser/provenance.py

```diff
@@ -63,6 +65,8 @@ class DictWithProvenance(dict):
         provenance_dict = dict_with_provenance.get_provenance()
     """
 
+    leaf_id_provenance = {}
+
```

Comment:
A doc string would useful here. A mini suggestion, but of course you can probably word this more accurately than I can:
```suggestion
    leaf_id_provenance = {}
    """
    dict : 
        This is a class-level variable to hold IDs of objects in memory. It is
        at the class level because this information needs to be passed around
        between various instances
    """
```